uaclient/apt.py

```diff
@@ -22,36 +23,37 @@
 CA_CERTIFICATES_FILE = '/usr/sbin/update-ca-certificates'
 
 
-class InvalidAPTCredentialsError(RuntimeError):
-    """Raised when invalid token is provided for APT access"""
-    pass
-
-
-def valid_apt_credentials(repo_url, username, password):
+def assert_valid_apt_credentials(repo_url, username, password):
     """Validate apt credentials for a PPA.
 
     @param repo_url: private-ppa url path
     @param username: PPA login username.
     @param password: PPA login password or resource token.
 
-    @return: True if valid or unable to validate
+    @raises: UserFacingError for invalid credentials or timeout
+    @raises: ProcessExecutionError for unexpected errors
     """
     protocol, repo_path = repo_url.split('://')
     if not os.path.exists('/usr/lib/apt/apt-helper'):
-        return True   # Do not validate
+        return None
     try:
         util.subp(['/usr/lib/apt/apt-helper', 'download-file',
                    '%s://%s:%s@%s/ubuntu/pool/' % (
                        protocol, username, password, repo_path),
-                   '/tmp/uaclient-apt-test'],
-                  capture=False)  # Hide credentials from logs
-        return True
-    except util.ProcessExecutionError:
-        pass
+                   '/tmp/uaclient-apt-test'])
+    except util.ProcessExecutionError as e:
+        if e.exit_code == 100:
+            stderr = str(e.stderr)
+            if re.search(r'401\s+Unauthorized|HttpError401', stderr):
+                raise exceptions.UserFacingError(
+                    'Invalid APT credentials provided for %s' % repo_url)
+            elif re.search(r'Connection timed out', stderr):
+                raise exceptions.UserFacingError(
+                    'Timeout trying to access %s' % repo_url)
+        raise
```

Comment:
Should we raise a catch-all `UserFacingError` here, so that unexpected problems don't cause a traceback?

Reply:
good thought. Fixing.